In [74]:
# Import libraries here that you need for different processing steps
import nltk
import csv
import spacy
import pandas as pd

In [75]:
import os
os.getcwd()

'C:\\Users\\Precious Olalere\\Documents\\IS 567 Text Mining'

In [76]:
# Reading in the train dataset and seeing the shape and types of the data 
train = pd.read_csv("train.csv")
train.head()


Unnamed: 0        label  \
0           0        false   
1           1    half-true   
2           2  mostly-true   
3           3        false   
4           4    half-true   

                                                                                                                                       statement  \
0                                                             Says the Annies List political group supports third-trimester abortions on demand.   
1  When did the decline of coal start? It started when natural gas took off that started to begin in (President George W.) Bushs administration.   
2                                      Hillary Clinton agrees with John McCain "by voting to give George Bush the benefit of the doubt on Iran."   
3                                                                 Health care reform legislation is likely to mandate free sex change surgeries.   
4                                                                                         The economic turnaround started at the end of my term.   

                                topic         speaker               context  \
0                            abortion    dwayne-bohac  State representative   
1  energy,history,job-accomplishments  scott-surovell        State delegate   
2                      foreign-policy    barack-obama             President   
3                         health-care    blog-posting                   NaN   
4                        economy,jobs   charlie-crist                   NaN   

   location       party             resource  
0     Texas  republican             a mailer  
1  Virginia    democrat      a floor speech.  
2  Illinois    democrat               Denver  
3       NaN        none       a news release  
4   Florida    democrat  an interview on CNN

In [77]:
#number of instances 
train.shape

(10240, 9)

In [78]:
#list of columns
train.dtypes

Unnamed: 0     int64
label         object
statement     object
topic         object
speaker       object
context       object
location      object
party         object
resource      object
dtype: object

In [79]:
#Converting all text from the “statement” column to lowercase.
train["statement"] = train["statement"].str.lower()
train["statement"].head(5)

0                                                               says the annies list political group supports third-trimester abortions on demand.
1    when did the decline of coal start? it started when natural gas took off that started to begin in (president george w.) bushs administration.
2                                        hillary clinton agrees with john mccain "by voting to give george bush the benefit of the doubt on iran."
3                                                                   health care reform legislation is likely to mandate free sex change surgeries.
4                                                                                           the economic turnaround started at the end of my term.
Name: statement, dtype: object

In [80]:
train.statement.tolist()[:10]

['says the annies list political group supports third-trimester abortions on demand.',
 'when did the decline of coal start? it started when natural gas took off that started to begin in (president george w.) bushs administration.',
 'hillary clinton agrees with john mccain "by voting to give george bush the benefit of the doubt on iran."',
 'health care reform legislation is likely to mandate free sex change surgeries.',
 'the economic turnaround started at the end of my term.',
 'the chicago bears have had more starting quarterbacks in the last 10 years than the total number of tenured (uw) faculty fired during the last two decades.',
 'jim dunnam has not lived in the district he represents for years now.',
 "i'm the only person on this stage who has worked actively just last year passing, along with russ feingold, some of the toughest ethics reform since watergate.",
 'however, it took $19.5 million in oregon lottery funds for the port of newport to eventually land the new noaa mari

In [82]:
#Extracting the quoted content from the statements and creating a new column “quoted_content” to contain the extracted output.

train["quoted_content"] = train["statement"].str.extract(r'"([^"]*)"')

#Checking the first 10 rows to verify quoted_ content column
train[["statement", "quoted_content"]].head(10)

statement  \
0                                                                               says the annies list political group supports third-trimester abortions on demand.   
1                    when did the decline of coal start? it started when natural gas took off that started to begin in (president george w.) bushs administration.   
2                                                        hillary clinton agrees with john mccain "by voting to give george bush the benefit of the doubt on iran."   
3                                                                                   health care reform legislation is likely to mandate free sex change surgeries.   
4                                                                                                           the economic turnaround started at the end of my term.   
5      the chicago bears have had more starting quarterbacks in the last 10 years than the total number of tenured (uw) faculty fired during the last two decades.   
6                                                                                            jim dunnam has not lived in the district he represents for years now.   
7  i'm the only person on this stage who has worked actively just last year passing, along with russ feingold, some of the toughest ethics reform since watergate.   
8                 however, it took $19.5 million in oregon lottery funds for the port of newport to eventually land the new noaa marine operations center-pacific.   
9                             says gop primary opponents glenn grothman and joe leibham cast a compromise vote that cost $788 million in higher electricity costs.   

                                                    quoted_content  
0                                                              NaN  
1                                                              NaN  
2  by voting to give george bush the benefit of the doubt on iran.  
3                                                              NaN  
4                                                              NaN  
5                                                              NaN  
6                                                              NaN  
7                                                              NaN  
8                                                              NaN  
9                                                              NaN

In [83]:
train.quoted_content.tolist()

[nan,
 nan,
 'by voting to give george bush the benefit of the doubt on iran.',
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 "who's actually turned around a government economy.",
 'took a position and the supreme court ruled unanimously that she was wrong.',
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 'says that any police officer can stop anyone who appears to them to be reasonably suspicious of being an undocumented person.',
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 'virtually nothing to the middle class.',
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,


Here are some of the examples from the quoted content:

"by voting to give george bush the benefit of the doubt on iran."
"who's actually turned around a government economy.",
'took a position and the supreme court ruled unanimously that she was wrong.',

In [84]:
#Remove non-alphanumeric characters that appear in statement (e.g. ‘<’, ‘>’, ‘*’, ‘+’.). 
#Convert all text to lowercase. Create a new column “cleaned_statement” to contain the extracted output

import re

train["cleaned_statement"] = train["statement"].str.lower().str.replace(r'[^a-zA-Z0-9\s]', '', regex=True)
train[["statement", "cleaned_statement"]].head(10)



statement  \
0                                                                               says the annies list political group supports third-trimester abortions on demand.   
1                    when did the decline of coal start? it started when natural gas took off that started to begin in (president george w.) bushs administration.   
2                                                        hillary clinton agrees with john mccain "by voting to give george bush the benefit of the doubt on iran."   
3                                                                                   health care reform legislation is likely to mandate free sex change surgeries.   
4                                                                                                           the economic turnaround started at the end of my term.   
5      the chicago bears have had more starting quarterbacks in the last 10 years than the total number of tenured (uw) faculty fired during the last two decades.   
6                                                                                            jim dunnam has not lived in the district he represents for years now.   
7  i'm the only person on this stage who has worked actively just last year passing, along with russ feingold, some of the toughest ethics reform since watergate.   
8                 however, it took $19.5 million in oregon lottery funds for the port of newport to eventually land the new noaa marine operations center-pacific.   
9                             says gop primary opponents glenn grothman and joe leibham cast a compromise vote that cost $788 million in higher electricity costs.   

                                                                                                                                             cleaned_statement  
0                                                                             says the annies list political group supports thirdtrimester abortions on demand  
1                     when did the decline of coal start it started when natural gas took off that started to begin in president george w bushs administration  
2                                                       hillary clinton agrees with john mccain by voting to give george bush the benefit of the doubt on iran  
3                                                                                health care reform legislation is likely to mandate free sex change surgeries  
4                                                                                                        the economic turnaround started at the end of my term  
5     the chicago bears have had more starting quarterbacks in the last 10 years than the total number of tenured uw faculty fired during the last two decades  
6                                                                                         jim dunnam has not lived in the district he represents for years now  
7  im the only person on this stage who has worked actively just last year passing along with russ feingold some of the toughest ethics reform since watergate  
8                  however it took 195 million in oregon lottery funds for the port of newport to eventually land the new noaa marine operations centerpacific  
9                           says gop primary opponents glenn grothman and joe leibham cast a compromise vote that cost 788 million in higher electricity costs

Before(line 1): when did the decline of coal start? it started when natural gas took off that started to begin in (president george w.) bushs administration.
After: when did the decline of coal start it started when natural gas took off that started to begin in president george w bushs administration

Before (line 5): the chicago bears have had more starting quarterbacks in the last 10 years than the total number of tenured (uw) faculty fired during the last two decades.
After: the chicago bears have had more starting quarterbacks in the last 10 years than the total number of tenured uw faculty fired during the last two decades

Before (line 9):says gop primary opponents glenn grothman and joe leibham cast a compromise vote that cost $788 million in higher electricity costs.
After:says gop primary opponents glenn grothman and joe leibham cast a compromise vote that cost 788 million in higher electricity costs

In [85]:
#Convert all text to lowercase
train = train.apply(lambda col: col.str.lower() if col.dtype == "object" else col)
train.head(10)


Unnamed: 0        label  \
0           0        false   
1           1    half-true   
2           2  mostly-true   
3           3        false   
4           4    half-true   
5           5         true   
6           6  barely-true   
7           7    half-true   
8           8    half-true   
9           9  mostly-true   

                                                                                                                                                         statement  \
0                                                                               says the annies list political group supports third-trimester abortions on demand.   
1                    when did the decline of coal start? it started when natural gas took off that started to begin in (president george w.) bushs administration.   
2                                                        hillary clinton agrees with john mccain "by voting to give george bush the benefit of the doubt on iran."   
3                                                                                   health care reform legislation is likely to mandate free sex change surgeries.   
4                                                                                                           the economic turnaround started at the end of my term.   
5      the chicago bears have had more starting quarterbacks in the last 10 years than the total number of tenured (uw) faculty fired during the last two decades.   
6                                                                                            jim dunnam has not lived in the district he represents for years now.   
7  i'm the only person on this stage who has worked actively just last year passing, along with russ feingold, some of the toughest ethics reform since watergate.   
8                 however, it took $19.5 million in oregon lottery funds for the port of newport to eventually land the new noaa marine operations center-pacific.   
9                             says gop primary opponents glenn grothman and joe leibham cast a compromise vote that cost $788 million in higher electricity costs.   

                                       topic                 speaker  \
0                                   abortion            dwayne-bohac   
1         energy,history,job-accomplishments          scott-surovell   
2                             foreign-policy            barack-obama   
3                                health-care            blog-posting   
4                               economy,jobs           charlie-crist   
5                                  education               robin-vos   
6                       candidates-biography  republican-party-texas   
7                                     ethics            barack-obama   
8                                       jobs          oregon-lottery   
9  energy,message-machine-2014,voting-record           duey-stroebel   

                      context   location         party  \
0        state representative      texas    republican   
1              state delegate   virginia      democrat   
2                   president   illinois      democrat   
3                         NaN        NaN          none   
4                         NaN    florida      democrat   
5  wisconsin assembly speaker  wisconsin    republican   
6                         NaN      texas    republican   
7                   president   illinois      democrat   
8                         NaN        NaN  organization   
9        state representative  wisconsin    republican   

                                   resource  \
0                                  a mailer   
1                           a floor speech.   
2                                    denver   
3                            a news release   
4                       an interview on cnn   
5                 a an online opinion-piece   
6                          a press release.   
7  a democratic debate in philadelphi

In [86]:
#list of columns
train.dtypes

Unnamed: 0            int64
label                object
statement            object
topic                object
speaker              object
context              object
location             object
party                object
resource             object
quoted_content       object
cleaned_statement    object
dtype: object

In [87]:
#Using NLTK to tokenize the statement text (statement & cleaned_statement column). 
#nltk.download('punkt')
from nltk.tokenize import word_tokenize
train["tokenized_statement"] = train["statement"].apply(word_tokenize)
train["tokenized_cleaned_statement"] = train["cleaned_statement"].apply(word_tokenize)
train[["statement","tokenized_statement","cleaned_statement","tokenized_cleaned_statement"]].head(10)


statement  \
0                                                                               says the annies list political group supports third-trimester abortions on demand.   
1                    when did the decline of coal start? it started when natural gas took off that started to begin in (president george w.) bushs administration.   
2                                                        hillary clinton agrees with john mccain "by voting to give george bush the benefit of the doubt on iran."   
3                                                                                   health care reform legislation is likely to mandate free sex change surgeries.   
4                                                                                                           the economic turnaround started at the end of my term.   
5      the chicago bears have had more starting quarterbacks in the last 10 years than the total number of tenured (uw) faculty fired during the last two decades.   
6                                                                                            jim dunnam has not lived in the district he represents for years now.   
7  i'm the only person on this stage who has worked actively just last year passing, along with russ feingold, some of the toughest ethics reform since watergate.   
8                 however, it took $19.5 million in oregon lottery funds for the port of newport to eventually land the new noaa marine operations center-pacific.   
9                             says gop primary opponents glenn grothman and joe leibham cast a compromise vote that cost $788 million in higher electricity costs.   

                                                                                                                                                                                   tokenized_statement  \
0                                                                                                     [says, the, annies, list, political, group, supports, third-trimester, abortions, on, demand, .]   
1                       [when, did, the, decline, of, coal, start, ?, it, started, when, natural, gas, took, off, that, started, to, begin, in, (, president, george, w., ), bushs, administration, .]   
2                                                                [hillary, clinton, agrees, with, john, mccain, ``, by, voting, to, give, george, bush, the, benefit, of, the, doubt, on, iran, ., '']   
3                                                                                                        [health, care, reform, legislation, is, likely, to, mandate, free, sex, change, surgeries, .]   
4                                                                                                                                  [the, economic, turnaround, started, at, the, end, of, my, term, .]   
5        [the, chicago, bears, have, had, more, starting, quarterbacks, in, the, last, 10, years, than, the, total, number, of, tenured, (, uw, ), faculty, fired, during, the, last, two, decades, .]   
6                                                                                                                [jim, dunnam, has, not, lived, in, the, district, he, represents, for, years, now, .]   
7  [i, 'm, the, only, person, on, this, stage, who, has, worked, actively, just, last, year, passing, ,, along, with, russ, feingold, ,, some, of, the, toughest, ethics, reform, since, watergate, .]   
8                       [however, ,, it, took, $, 19.5, million, in, oregon, lottery, funds, for, the, port, of, newport, to, eventually, land, the, new, noaa, marine, operations, center-pacific, .]   
9                                       [says, gop, primary, opponents, glenn, grothman, and, joe, leibham, cast, a, compromise, vote, that, cost, $, 788, million, in, higher, electricity, costs, .]   

                                                                                                       

#Describe the differences between tokenized_statement and tokenized_cleaned_statement. 

One of the major difference in the tokenized_statement and tokenized_cleaned_statement is how each words are seperated. From the tokenized statement, it is clear that not only are the words seperated by a comma but even non-alphanumeric characters such as ‘(‘, ')','.' are seperated by a comma, whereas in the tokenized_cleaned_statement only the words are seperated by a comma. Below is also an example of tokenized_statement and tokenized_cleaned_statement from line 8:

Tokenized_statement:[however, ,, it, took, $, 19.5, million, in, oregon, lottery, funds, for, the, port, of, newport, to, eventually, land, the, new, noaa, marine, operations, center-pacific, .]

Tokenized_cleaned_statement:[however, it, took, 195, million, in, oregon, lottery, funds, for, the, port, of, newport, to, eventually, land, the, new, noaa, marine, operations, centerpacific]

In [88]:
#Using another package (e.g., spacy) or your own code to tokenize the statement text.

nlp = spacy.load('en_core_web_sm')
def tokenize_spacy(text):
    # Process the text with spaCy
    doc = nlp(text)
    # Extract tokens
    return [token.text for token in doc]

train['tokenized_statement_spacy'] = train['statement'].apply(tokenize_spacy)


In [89]:
train['tokenized_cleaned_statement_spacy'] = train['cleaned_statement'].apply(tokenize_spacy)


In [90]:
train[['statement', 'tokenized_statement_spacy', 'cleaned_statement', 'tokenized_cleaned_statement_spacy']].head(10)


statement  \
0                                                                               says the annies list political group supports third-trimester abortions on demand.   
1                    when did the decline of coal start? it started when natural gas took off that started to begin in (president george w.) bushs administration.   
2                                                        hillary clinton agrees with john mccain "by voting to give george bush the benefit of the doubt on iran."   
3                                                                                   health care reform legislation is likely to mandate free sex change surgeries.   
4                                                                                                           the economic turnaround started at the end of my term.   
5      the chicago bears have had more starting quarterbacks in the last 10 years than the total number of tenured (uw) faculty fired during the last two decades.   
6                                                                                            jim dunnam has not lived in the district he represents for years now.   
7  i'm the only person on this stage who has worked actively just last year passing, along with russ feingold, some of the toughest ethics reform since watergate.   
8                 however, it took $19.5 million in oregon lottery funds for the port of newport to eventually land the new noaa marine operations center-pacific.   
9                             says gop primary opponents glenn grothman and joe leibham cast a compromise vote that cost $788 million in higher electricity costs.   

                                                                                                                                                                             tokenized_statement_spacy  \
0                                                                                                 [says, the, annies, list, political, group, supports, third, -, trimester, abortions, on, demand, .]   
1                       [when, did, the, decline, of, coal, start, ?, it, started, when, natural, gas, took, off, that, started, to, begin, in, (, president, george, w., ), bushs, administration, .]   
2                                                                  [hillary, clinton, agrees, with, john, mccain, ", by, voting, to, give, george, bush, the, benefit, of, the, doubt, on, iran, ., "]   
3                                                                                                        [health, care, reform, legislation, is, likely, to, mandate, free, sex, change, surgeries, .]   
4                                                                                                                                  [the, economic, turnaround, started, at, the, end, of, my, term, .]   
5        [the, chicago, bears, have, had, more, starting, quarterbacks, in, the, last, 10, years, than, the, total, number, of, tenured, (, uw, ), faculty, fired, during, the, last, two, decades, .]   
6                                                                                                                [jim, dunnam, has, not, lived, in, the, district, he, represents, for, years, now, .]   
7  [i, 'm, the, only, person, on, this, stage, who, has, worked, actively, just, last, year, passing, ,, along, with, russ, feingold, ,, some, of, the, toughest, ethics, reform, since, watergate, .]   
8                   [however, ,, it, took, $, 19.5, million, in, oregon, lottery, funds, for, the, port, of, newport, to, eventually, land, the, new, noaa, marine, operations, center, -, pacific, .]   
9                                       [says, gop, primary, opponents, glenn, grothman, and, joe, leibham, cast, a, compromise, vote, that, cost, $, 788, million, in, higher, electricity, costs, .]   

                                                                                                       

#After working with both tokenization methods, write about their similarities and differences using samples from “tokenized_statement” column. Do you observe that one works better than the other for this dataset? If so, explain your observation and proceed with the tokenizer that performs better. 

One of the major similarities between NLTK and Spacy is that it seperates each words with a comma, including the non-alphanumeric characters, however, the major difference is how spacy seperates words with an hyphen into seperate words. Here is an example from line 0 

NLTK tokenized_statement:[says, the, annies, list, political, group, supports, third-trimester, abortions, on, demand, .]

spacy tokenized_statement:[says, the, annies, list, political, group, supports, third, -, trimester, abortions, on, demand, .]

The other difference between NLTK and Spacy tokenization is how Spacy changed the meaning of the word by seperating one word into two words. 

Also I realized that Spacy took a while to process the text compared to when I used NLTK. Based on all the facts provided above, I am inclined to say that NLTK works better. Hence I will be progressing with the NLTK tokenised statement. 

In [91]:
#Using NLTK to remove stopwords from the tokenized_cleaned_statement and create column named stopword_removed_statement.
nltk.download("stopwords")


[nltk_data] Downloading package stopwords to C:\Users\Precious
[nltk_data]     Olalere\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [92]:
from nltk.corpus import stopwords

stop_words = set(stopwords.words("english"))
def remove_stopwords(tokenized_text):
    return [word for word in tokenized_text if word not in stop_words]

train["stopword_removed_statement"] = train["tokenized_cleaned_statement"].apply(remove_stopwords)
train[["tokenized_cleaned_statement", "stopword_removed_statement"]].head(10)


tokenized_cleaned_statement  \
0                                                                                             [says, the, annies, list, political, group, supports, thirdtrimester, abortions, on, demand]   
1                        [when, did, the, decline, of, coal, start, it, started, when, natural, gas, took, off, that, started, to, begin, in, president, george, w, bushs, administration]   
2                                                               [hillary, clinton, agrees, with, john, mccain, by, voting, to, give, george, bush, the, benefit, of, the, doubt, on, iran]   
3                                                                                               [health, care, reform, legislation, is, likely, to, mandate, free, sex, change, surgeries]   
4                                                                                                                         [the, economic, turnaround, started, at, the, end, of, my, term]   
5     [the, chicago, bears, have, had, more, starting, quarterbacks, in, the, last, 10, years, than, the, total, number, of, tenured, uw, faculty, fired, during, the, last, two, decades]   
6                                                                                                       [jim, dunnam, has, not, lived, in, the, district, he, represents, for, years, now]   
7  [im, the, only, person, on, this, stage, who, has, worked, actively, just, last, year, passing, along, with, russ, feingold, some, of, the, toughest, ethics, reform, since, watergate]   
8                      [however, it, took, 195, million, in, oregon, lottery, funds, for, the, port, of, newport, to, eventually, land, the, new, noaa, marine, operations, centerpacific]   
9                                 [says, gop, primary, opponents, glenn, grothman, and, joe, leibham, cast, a, compromise, vote, that, cost, 788, million, in, higher, electricity, costs]   

                                                                                                               stopword_removed_statement  
0                                                     [says, annies, list, political, group, supports, thirdtrimester, abortions, demand]  
1                        [decline, coal, start, started, natural, gas, took, started, begin, president, george, w, bushs, administration]  
2                                              [hillary, clinton, agrees, john, mccain, voting, give, george, bush, benefit, doubt, iran]  
3                                                      [health, care, reform, legislation, likely, mandate, free, sex, change, surgeries]  
4                                                                                              [economic, turnaround, started, end, term]  
5               [chicago, bears, starting, quarterbacks, last, 10, years, total, number, tenured, uw, faculty, fired, last, two, decades]  
6                                                                                       [jim, dunnam, lived, district, represents, years]  
7           [im, person, stage, worked, actively, last, year, passing, along, russ, feingold, toughest, ethics, reform, since, watergate]  
8    [however, took, 195, million, oregon, lottery, funds, port, newport, eventually, land, new, noaa, marine, operations, centerpacific]  
9  [says, gop, primary, opponents, glenn, grothman, joe, leibham, cast, compromise, vote, cost, 788, million, higher, electricity, costs]

In [93]:
#Using Porter Stemmer (implemented in NLTK) to stem the text in stopword_removed_statement and 
#create a new column called stemmed_statement.

from nltk.stem import PorterStemmer


In [94]:
stemmer = PorterStemmer()

def stem_words(tokenized_text):
    return [stemmer.stem(word) for word in tokenized_text]
train["stemmed_statement"] = train["stopword_removed_statement"].apply(stem_words)
train[["stopword_removed_statement", "stemmed_statement"]].head(10)


stopword_removed_statement  \
0                                                     [says, annies, list, political, group, supports, thirdtrimester, abortions, demand]   
1                        [decline, coal, start, started, natural, gas, took, started, begin, president, george, w, bushs, administration]   
2                                              [hillary, clinton, agrees, john, mccain, voting, give, george, bush, benefit, doubt, iran]   
3                                                      [health, care, reform, legislation, likely, mandate, free, sex, change, surgeries]   
4                                                                                              [economic, turnaround, started, end, term]   
5               [chicago, bears, starting, quarterbacks, last, 10, years, total, number, tenured, uw, faculty, fired, last, two, decades]   
6                                                                                       [jim, dunnam, lived, district, represents, years]   
7           [im, person, stage, worked, actively, last, year, passing, along, russ, feingold, toughest, ethics, reform, since, watergate]   
8    [however, took, 195, million, oregon, lottery, funds, port, newport, eventually, land, new, noaa, marine, operations, centerpacific]   
9  [says, gop, primary, opponents, glenn, grothman, joe, leibham, cast, compromise, vote, cost, 788, million, higher, electricity, costs]   

                                                                                                            stemmed_statement  
0                                                       [say, anni, list, polit, group, support, thirdtrimest, abort, demand]  
1                              [declin, coal, start, start, natur, ga, took, start, begin, presid, georg, w, bush, administr]  
2                                       [hillari, clinton, agre, john, mccain, vote, give, georg, bush, benefit, doubt, iran]  
3                                                     [health, care, reform, legisl, like, mandat, free, sex, chang, surgeri]  
4                                                                                      [econom, turnaround, start, end, term]  
5              [chicago, bear, start, quarterback, last, 10, year, total, number, tenur, uw, faculti, fire, last, two, decad]  
6                                                                                 [jim, dunnam, live, district, repres, year]  
7            [im, person, stage, work, activ, last, year, pass, along, russ, feingold, toughest, ethic, reform, sinc, waterg]  
8        [howev, took, 195, million, oregon, lotteri, fund, port, newport, eventu, land, new, noaa, marin, oper, centerpacif]  
9  [say, gop, primari, oppon, glenn, grothman, joe, leibham, cast, compromis, vote, cost, 788, million, higher, electr, cost]

#Report 3 examples before and after stemming.

Here are 3 examples before and after applying Porter Stemmer to the stopword_removed_statement

Before (line 1): [decline, coal, start, started, natural, gas, took, started, begin, president, george, w, bushs, administration]

After:[declin, coal, start, start, natur, ga, took, start, begin, presid, georg, w, bush, administr]

Before (line 3) : [health, care, reform, legislation, likely, mandate, free, sex, change, surgeries]	

After(line 4) : [health, care, reform, legisl, like, mandat, free, sex, chang, surgeri]

Before (line 7): [im, person, stage, worked, actively, last, year, passing, along, russ, feingold, toughest, ethics, reform, since, watergate]
After: [im, person, stage, work, activ, last, year, pass, along, russ, feingold, toughest, ethic, reform, sinc, waterg]

In [95]:
#Using another stemmer included in NLTK to stem the stopword_removed_statement text. 
#Write about their similarities and differences.
#Using the snowball stemmer

from nltk.stem import SnowballStemmer


In [96]:
# Importing the Snowball Stemmer for English
snowball_stemmer = SnowballStemmer("english")

def snowball_stem_words(tokenized_text):
    return [snowball_stemmer.stem(word) for word in tokenized_text]

train["snowball_stemmed_statement"] = train["stopword_removed_statement"].apply(snowball_stem_words)

train[["stopword_removed_statement", "snowball_stemmed_statement"]].head(10)


stopword_removed_statement  \
0                                                     [says, annies, list, political, group, supports, thirdtrimester, abortions, demand]   
1                        [decline, coal, start, started, natural, gas, took, started, begin, president, george, w, bushs, administration]   
2                                              [hillary, clinton, agrees, john, mccain, voting, give, george, bush, benefit, doubt, iran]   
3                                                      [health, care, reform, legislation, likely, mandate, free, sex, change, surgeries]   
4                                                                                              [economic, turnaround, started, end, term]   
5               [chicago, bears, starting, quarterbacks, last, 10, years, total, number, tenured, uw, faculty, fired, last, two, decades]   
6                                                                                       [jim, dunnam, lived, district, represents, years]   
7           [im, person, stage, worked, actively, last, year, passing, along, russ, feingold, toughest, ethics, reform, since, watergate]   
8    [however, took, 195, million, oregon, lottery, funds, port, newport, eventually, land, new, noaa, marine, operations, centerpacific]   
9  [says, gop, primary, opponents, glenn, grothman, joe, leibham, cast, compromise, vote, cost, 788, million, higher, electricity, costs]   

                                                                                                   snowball_stemmed_statement  
0                                                       [say, anni, list, polit, group, support, thirdtrimest, abort, demand]  
1                             [declin, coal, start, start, natur, gas, took, start, begin, presid, georg, w, bush, administr]  
2                                       [hillari, clinton, agre, john, mccain, vote, give, georg, bush, benefit, doubt, iran]  
3                                                     [health, care, reform, legisl, like, mandat, free, sex, chang, surgeri]  
4                                                                                      [econom, turnaround, start, end, term]  
5              [chicago, bear, start, quarterback, last, 10, year, total, number, tenur, uw, faculti, fire, last, two, decad]  
6                                                                                 [jim, dunnam, live, district, repres, year]  
7            [im, person, stage, work, activ, last, year, pass, along, russ, feingold, toughest, ethic, reform, sinc, waterg]  
8        [howev, took, 195, million, oregon, lotteri, fund, port, newport, eventu, land, new, noaa, marin, oper, centerpacif]  
9  [say, gop, primari, oppon, glenn, grothman, joe, leibham, cast, compromis, vote, cost, 788, million, higher, electr, cost]

In [97]:
#Displaying Porter stemmed statement and snowball stemmed statement

train[["stemmed_statement", "snowball_stemmed_statement"]].head(10)


stemmed_statement  \
0                                                       [say, anni, list, polit, group, support, thirdtrimest, abort, demand]   
1                              [declin, coal, start, start, natur, ga, took, start, begin, presid, georg, w, bush, administr]   
2                                       [hillari, clinton, agre, john, mccain, vote, give, georg, bush, benefit, doubt, iran]   
3                                                     [health, care, reform, legisl, like, mandat, free, sex, chang, surgeri]   
4                                                                                      [econom, turnaround, start, end, term]   
5              [chicago, bear, start, quarterback, last, 10, year, total, number, tenur, uw, faculti, fire, last, two, decad]   
6                                                                                 [jim, dunnam, live, district, repres, year]   
7            [im, person, stage, work, activ, last, year, pass, along, russ, feingold, toughest, ethic, reform, sinc, waterg]   
8        [howev, took, 195, million, oregon, lotteri, fund, port, newport, eventu, land, new, noaa, marin, oper, centerpacif]   
9  [say, gop, primari, oppon, glenn, grothman, joe, leibham, cast, compromis, vote, cost, 788, million, higher, electr, cost]   

                                                                                                   snowball_stemmed_statement  
0                                                       [say, anni, list, polit, group, support, thirdtrimest, abort, demand]  
1                             [declin, coal, start, start, natur, gas, took, start, begin, presid, georg, w, bush, administr]  
2                                       [hillari, clinton, agre, john, mccain, vote, give, georg, bush, benefit, doubt, iran]  
3                                                     [health, care, reform, legisl, like, mandat, free, sex, chang, surgeri]  
4                                                                                      [econom, turnaround, start, end, term]  
5              [chicago, bear, start, quarterback, last, 10, year, total, number, tenur, uw, faculti, fire, last, two, decad]  
6                                                                                 [jim, dunnam, live, district, repres, year]  
7            [im, person, stage, work, activ, last, year, pass, along, russ, feingold, toughest, ethic, reform, sinc, waterg]  
8        [howev, took, 195, million, oregon, lotteri, fund, port, newport, eventu, land, new, noaa, marin, oper, centerpacif]  
9  [say, gop, primari, oppon, glenn, grothman, joe, leibham, cast, compromis, vote, cost, 788, million, higher, electr, cost]

I did not really see a major difference between the porter stemmer and the snowball stemmer, comparing both result, they both had the same result 

In [99]:
#Checking for any other observations on the data preprocessing process
train.dtypes

Unnamed: 0                            int64
label                                object
statement                            object
topic                                object
speaker                              object
context                              object
location                             object
party                                object
resource                             object
quoted_content                       object
cleaned_statement                    object
tokenized_statement                  object
tokenized_cleaned_statement          object
tokenized_statement_spacy            object
tokenized_cleaned_statement_spacy    object
stopword_removed_statement           object
stemmed_statement                    object
snowball_stemmed_statement           object
dtype: object

In [100]:
#Checking for any other observations on the data preprocessing process.
#Checking for missing data

train.isnull().sum()

Unnamed: 0                              0
label                                   0
statement                               0
topic                                   2
speaker                                 2
context                              2897
location                             2208
party                                   2
resource                              102
quoted_content                       9843
cleaned_statement                       0
tokenized_statement                     0
tokenized_cleaned_statement             0
tokenized_statement_spacy               0
tokenized_cleaned_statement_spacy       0
stopword_removed_statement              0
stemmed_statement                       0
snowball_stemmed_statement              0
dtype: int64

Write-up
•	In a short paragraph, discuss any other observations on the dataset.
In terms of any other observations, the dataset seems to be textual data about political campaigns, perhaps speech given during a political campaign by different parties. The dataset appears to have some missing values in some columns, with a large portion missing items coming from the context and location column. Due to the fact that majority of the sentences do not have a quoted content, there is also about 9843 missing items from the new column that I created after I extracted the quoted content from the statement column.
•	In a short paragraph, discuss any other observations on the data preprocessing process.
As I progressed with different preprocessing processes, I realized that the meaning of each statement started to change, and it changed drastically after I performed stop words remover and stemming. These two processes literally changed the text. For example, the stop word remover extracted only useful information from the text and performing stemming reduced the plain text words to their root word
•	Besides the preprocessing steps that you have taken above, are there other things that can be done to improve the “cleanliness” of the data and can aid in text mining? List as many as possible! 
Other forms of preprocessing steps that can be done to improve the cleanliness of the data includes; lemmatization, spell checking, tokenization and stop word removal for the resource column might also be helpful.


